File: sol/872.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <algorithm>

struct Solution {
    void leafSeq(TreeNode* root, vector<int> &res){
        if (root == nullptr){
            return;
        }
        if (root->left == nullptr && root->right == nullptr){
            res.push_back(root->val);
            return;
        }
        this->leafSeq(root->left, res);
        this->leafSeq(root->right, res);
    }
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<int> res1, res2;
        this->leafSeq(root1, res1);
        this->leafSeq(root2, res2);
        return equal(res1.begin(), res1.end(), res2.begin(), res2.end());
    }
};

File: sol/872/872.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        def leafSeq(node):
            if node is None:
                return
            if node.left is None and node.right is None:
                yield node.val
            else:
                yield from leafSeq(node.left)
                yield from leafSeq(node.right)

        return tuple(leafSeq(root1)) == tuple(leafSeq(root2))
